plugin/core/views.py

```diff
@@ -172,3 +174,22 @@ def text_document_range_formatting(view: sublime.View, region: sublime.Region) -
         "options": formatting_options(view.settings()),
         "range": region_to_range(view, region).to_lsp()
     })
+
+
+def minihtml(content: Union[str, dict], view: sublime.View) -> str:
+    """ Content can be a string or MarkupContent """
+    if isinstance(content, str):
+        return html.escape(content)
```

Comment:
In `pyls` I get this completion item for `json.JSONDecoreError`:
```
    {
      "kind": 7,
      "documentation": "JSONDecodeError(msg: str, doc: str, pos: int)\n\nSubclass of ValueError with the following additional properties:\n\nmsg: The unformatted error message\ndoc: The JSON document being parsed\npos: The start index of doc where parsing failed\nlineno: The line corresponding to pos\ncolno: The column corresponding to pos",
      "sortText": "aJSONDecodeError",
      "insertText": "JSONDecodeError",
      "detail": "json",
      "label": "JSONDecodeError"
    },
```
and the documentation is rendered without line breaks:
![Screenshot 2020-05-08 at 10 28 13](https://user-images.githubusercontent.com/153197/81387444-fc836900-9116-11ea-926a-e2fc40a5a0a0.png)

Which makes it quite hard to parse.

Do we need to convert line breaks to HTML or something?
Should we do this instead:
```py
return mdpopups.md2html(view, html.escape(content))
```
 (and same for escape below)

Reply:
If it is plaintext, then:

1. Escape it with html.escape
2. The whitespace should be rendered verbatim. If we put the HTML-escaped plaintext string into ST's minihtml renderer it removes extraneous whitespace (like a browser does too).

We account for (1) with `html.escape`, but you're right that we don't account for (2).

Perhaps we can simply wrap the HTML-escaped plaintext in a `<code></code>` block?

Reply:
Actually there is also another issue with using `html.escape` that needs to be investigated.
The `<http://json.org>` should be shown verbatim. The angle brackets turn into entities.
![Screenshot 2020-05-08 at 10 45 24](https://user-images.githubusercontent.com/153197/81388947-56852e00-9119-11ea-9d33-f292a3225fbc.png)

Reply:
Yes, so perhaps wrap it in `<code>` or `<tt>`. https://www.sublimetext.com/docs/3/minihtml.html

Reply:
Meaning `"<code>{}</code>".format(html.escape(value))`

Reply:
We can't `html.escape` as that will mess up the angle brackets from the example above.
But we need to experiment more and definitively add tests for that as it's way too easy to break.

Reply:
But `&lt;` and `&gt;` are supposed to be rendered as `<` and `>` in a `<code>` tag?

```
<code>
&lgt;code&gt;
</code>
```
Is supposed to be rendered as
```
<code>
```

Reply:
It doesn't seem to work but I don't have time to investigate more right now. Maybe I was testing wrong...

Reply:
When I run this in the console:

```
>>> view.show_popup('<code>&lt;code&gt;</code>')
```
I get the correct popup. So I'm pretty sure this will give the correct output:

```
return '<code>{}</code>'.format(html.escape(content))
```

Reply:
It would be nice if minihtml supported the `<pre>` tag, but oh well.

Reply:
But line breaks are not preserved:
```
view.show_popup('<code>&lt;code&gt;\n\nEND</code>')
```

Reply:
Oh, I didn't know that :) I guess we really do need a `<pre>` tag. Man, that is so annoying.

I guess you were right about using mdpopups.
```
Python 3.3>>> mdpopups.show_popup(view, '```\nfoo\nbar\n```', md=True)
```
It seems to take care of linebreaks and whitespace-preservation when you wrap the plaintext in three backticks:
```
return mdpopups.md2html(view, "```\n{}\n```".format(content))
```

Reply:
Well, if the content happens to contain a line with three backticks, this could break. But I think that's a problem for later.

Reply:
It's ironic that markdown is an easier problem to solve than plaintext...

Reply:
It's not the problem that minihtml doesn't support `pre` tag but that it doesn't support `white-space: pre`. `code` tag should also have that by default per spec.

Reply:
https://forum.sublimetext.com/t/does-minihtml-support-the-pre-tag/46267/5
instead of the pre tag, we can replace the `\n` with `<br>` to **pre**serve the whitespace.

Reply:
You would still need to account for extra space characters starting at a line (for indentation for example). Let’s not reinvent the wheel and just use mdpopups.

Reply:
I been thinking. We cannot apply md2html to markupKind "plaintext". 
Maybe the plaintext contain a `#` as the first char which would then accidentally get treated as h1  :)

Also I now encountered a bug, where sometimes the pyls return this:
{'insertText': 'Any', 'sortText': 'aAny', 'kind': 7, 'detail': 'plugin.completion', 'label': 'Any', 'documentation': 'Any()\n------------------------------\n'}

![a](https://user-images.githubusercontent.com/22029477/81440248-008fa500-9170-11ea-9712-c6ee3af5fb10.png)

and sometimes it returns this:
{'detail': 'plugin.completion', 'documentation': 'Any()', 'kind': None, 'label': 'Any', 'sortText': 'aAny', 'insertText': 'Any'}

![b](https://user-images.githubusercontent.com/22029477/81440260-071e1c80-9170-11ea-9a4f-ebc27b855ab9.png)

Reply:
So we should also `strip()` the text?

Reply:
My proposal above avoids all this.

Reply:
I’m on a phone right now but it’s this:

```
return mdpopups.md2html(view, "```\n{}\n```".format(content))
```

Again no need to use HTML escape

Reply:
here is 
`html.escape(content).replace('\n', '<br>')`

![image](https://user-images.githubusercontent.com/22029477/81442273-97119580-9173-11ea-9273-987d1cd3bb11.png)

`mdpopups.md2html(view, "```\n{}\n```".format(content))`

![image](https://user-images.githubusercontent.com/22029477/81442605-33d43300-9174-11ea-94c9-7601eec204ba.png)

Reply:
The last one is more faithful.

When there is whitespace like this:

```
some docs

    here is some indentation!
```

then that will be rendered correctly by your last example.

Reply:
> BTW. ST closes completion list when it can't fit the "cooperate" popup...

Yes this happen to me now when I show docs for JSONDecodeError, but it doesn't happen always.